give a better error message when a remote ref no longer exists

doy opened the pull request:
if we can't find the given ref after fetching, try clearing the cache
ourselves and trying again. if we still can't find it, we know that it
actually just doesn't exist, so tell the user that.

Fixes #4140.

Commit:
give a better error message when a remote ref no longer exists

if we can't find the given ref after fetching, try clearing the cache
ourselves and trying again. if we still can't find it, we know that it
actually just doesn't exist, so tell the user that.

segiddins commented on lib/bundler/source/git/git_proxy.rb:
```diff
@@ -27,6 +27,13 @@ def initialize(command, path = nil, extra_info = nil)
         end
       end
 
+      class GitRevisionError < GitError
```
`MissingGitRevisionError`

segiddins commented:
Sweet! Think you could throw in a few tests for this?

Commit:
more descriptive error class name

Commit:
add some tests for missing git refs

doy commented:
How does this look? I wanted to add some tests for an actual corrupted git repo cache, but I couldn't figure out how the caching stuff works in tests. This covers the major functionality though.

segiddins commented:
Looks good! I'd also like to see a test for removing the cache and retrying when the first attempt to find the revision fails

doy commented:
I did some more digging, and I can't actually find a situation where the first attempt would fail but the second attempt would succeed. Everything I tried led to either the cache being corrupted enough that it failed at a much earlier stage, or it worked fine the first time. Would it be reasonable to remove the "remove the cache and retry" logic? I'm not entirely convinced that I didn't miss anything (I'm not familiar with this codebase at all), but leaving it in "just in case" without being able to test it also doesn't sound ideal.

indirect commented:
I'm okay removing the remove cache option, and we can see how it goes.

Commit:
remove the "clear cache and retry" logic

i can't find a case where it will actually be hit, so this avoids
leaving untested code around

doy commented:
Done.

segiddins commented:
@homu r+

homu commented:
:pushpin: Commit 0cb5e26 has been approved by `segiddins`

<!-- @homu r=segiddins 0cb5e26f94a3e4aba241ab0d1dafa48db84fe175 -->

homu commented:
:hourglass: Testing commit 0cb5e26 with merge 76bfa53...

homu commented:
:boom: Test timed out

indirect commented:
@homu retry

homu commented:
:hourglass: Testing commit 0cb5e26 with merge 1d00f6e...

homu commented:
:broken_heart: Test failed - [status](https://travis-ci.org/bundler/bundler/builds/95748512)

doy commented:
Anything else I need to do here? It doesn't look like this failure is related to my branch.

segiddins commented:
@homu r+

homu commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.

homu commented:
:pushpin: Commit 0cb5e26 has been approved by `segiddins`

<!-- @homu r=segiddins 0cb5e26f94a3e4aba241ab0d1dafa48db84fe175 -->

homu commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.

segiddins commented:
@homu retry

homu commented:
:hourglass: Testing commit 0cb5e26 with merge 8755a1f...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/bundler/bundler/builds/97124446)
